util: add util.callbackify()

refack opened the pull request:
This is a first draft of an implementation of `util.callbackify(function)`.
There is at least one issue that need ironing out: if and how to project the return value into multiple args...

Implementation inspired by @addaleax's #12442
Background discussion at: https://github.com/nodejs/CTC/issues/109

## Open discussion
Should we allow customization of the wrapping?
  1. specifying a transform for the `awaitable` before the connection to the callback
  2. projection of the returned value
  3. tiying in the callback into the promise's life-cycle


---

#### Original commit descriptions

##### Add `util.callbackify(function)` for creating callback style functions
from functions returning a `Promise`/`awaitable`

Fixes: https://github.com/nodejs/CTC/issues/109



<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
util

refack commented:
For how to treat the returned value I choose to:
1. By default pass on as-is
2. Allow the user to ask us to splat the value
3. Allow the user to specify a list of keys that we will project out of the value
4. [Just thought of] allow the user to provide a projection lambda

refack commented:
/cc @chrisdickinson @benjamingr @Fishrock123 @nodejs/ctc

not-an-aardvark commented:
> For how to treat the returned value I choose to:
> 1. By default pass on as-is
> 2. Allow the user to ask us to splat the value
> 3. Allow the user to specify a list of keys that we will project out of the value
> 4. [Just thought of] allow the user to provide a projection lambda

I think option 1 is definitely best. The other options can easily be simulated in userland by using destructuring in the callback, so I don't think it's necessary to have special behavior to for spreading arrays/objects in core.

TimothyGu commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
+      throw new TypeError('The [util.callbackify.custom] property must be ' +
+        'a function');
+    }
+    Object.defineProperty(fn, kCustomCallbackifyiedSymbol, {
```
I'd protect this with `if (fn[kCustomCallbackifyiedSymbol] !== fn)` as `Object.defineProperty` may be quite slow.

TimothyGu commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
+      } else if (project) {
+        const allThere = project.all((k) => k in ret);
+        if (!allThere) {
+          throw new TypeError('Not all of the projected properties' +
```
I'd fill in `undefined` for properties that are not present.

not-an-aardvark replied:
If we keep this property-picking option, then I think this error should be sent with `cb(new TypeError(...))` rather than throwing and rejecting the Promise. If someone is using `callbackify`, then they're probably going to be listening for errors in the callback rather than listening for Promise rejections. (However, I don't think we should keep the property-picking option -- see https://github.com/nodejs/node/pull/12712#issuecomment-297895360.)

refack replied:
I was wondering about that too.
Can we find some precedents? `lodash` maybe?

refack replied:
Yep. Didn't think of that.

TimothyGu commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
+
+callbackify.custom = kCustomCallbackifyiedSymbol;
+callbackify.toSplat = kCustomCallbackifyiedToSplat;
+callbackify.project = kCustomCallbackifyiedProject;
```
We should not make `toSplat` or `project` public for now, just like for `promisify`.

refack replied:
🤔 not sure, but keeping in line with `promisify`

TimothyGu commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
 process.versions[exports.inspect.custom] =
   (depth) => exports.format(JSON.parse(JSON.stringify(process.versions)));
+
+const kCustomCallbackifyiedSymbol = Symbol('util.callbackify.custom');
```
```diff
-kCustomCallbackifyiedSymbol
+kCustomCallbackifiedSymbol 
```

?

vkurchatkin commented:
This seems unnecessarily complicated for such a simple feature. Is there an example where you would need all that customization?

not-an-aardvark commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
+      } else {
+        cb(null, ret);
+      }
+    }, cb);
```
This could cause confusing behavior if the original Promise rejects with a non-error:

```js
function promiseFunction() {
  return Promise.reject(null);
}

util.callbackify(promiseFunction)((err, result) => {
  // err will be null, which could be confusing because the promise actually rejected.
});
```

Bluebird seems to handle this by [wrapping the rejection reason in an `Error` if it's falsy](https://github.com/petkaantonov/bluebird/blob/2207fae3572f03b089bc92d3a6cefdd278cff7ab/src/nodeify.js#L30-L43).

refack replied:
Maybe even if `!err instanceof Error`?

not-an-aardvark replied:
I don't think that would be a good idea because people sometimes intentionally reject Promises with objects. If we wrapped the objects in errors then we would lose information about the properties.

refack replied:
Adapted BlueBird snippet

vkurchatkin commented on lib/util.js:
```diff
@@ -1057,3 +1057,68 @@ exports._exceptionWithHostPort = function(err,
+    }
+    orig.call(this, ...args).then((ret) => {
+      if (toSplat) {
+        cb(null, ...ret);
```
What if `cb` throws? We probably want to avoid getting rejected promises from callbackified functions

refack replied:
What can we do? who can we tell?
Maybe `process.emit('uncaughtException')` or 'unhandledRejection'?

vkurchatkin replied:
Maybe wrap the whole thing in `process.nextTick`?

refack replied:
So it'll go out of the Promise timeline?
I'd rather explicitly explode a la https://github.com/nodejs/node/pull/12712/files#diff-3deb3f32958bb937ae05c6f3e4abbdf5R1121

vkurchatkin replied:
Not a good idea, it's better to allow the error to bubble. Otherwise you'll have to duplicate a lot of logic here for no good reason.

refack replied:
I'll check it out... (don't want to duplicate!)

not-an-aardvark commented:
Is there a use case for having `util.callbackify.custom`? `util.promisify.custom` is useful because some functions have different callback conventions and so the default implementation wouldn't work, but that isn't the case with `util.callbackify` because all promise-returning functions just return a promise without separate conventions.

refack commented:
> Is there a use case for having util.callbackify.custom?

🤔 I thought of it as an 'escape hatch' to allow users to do whatever. Maybe adapt to old `Promise` implementations like our [mpromise](https://github.com/aheckmann/mpromise) that doesn't have a `catch`

I have in mind the use case where this is used by dev A to adapt dev B's and dev C's APIs into a new uniform API, and there is a need to tweak each adaptation a little. So I'm not even sure about hiding `callbackify.toSplat` && `callbackify.project`. I think it makes the implementation more complete...

not-an-aardvark commented:
I think having customization here is unnecessary given that it's trivial for users to customize the behavior in userland. Adding extra configurability where it's not needed seems like it will make the API bloated.

refack commented:
P.S. do you think we should return the Promise
`     return orig.call(this, ...args).then((ret) => {`
https://github.com/nodejs/node/pull/12712/files#diff-3deb3f32958bb937ae05c6f3e4abbdf5R1096
So the user could add a `.catch` or chain something?

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,81 @@ exports._exceptionWithHostPort = function(err,
+
+const kCustomCallbackifiedSymbol = Symbol('util.callbackify.custom');
+const kCustomCallbackifiedToSplat = Symbol('util.callbackify.toSplat');
+const kCustomCallbackifiedProject = Symbol('util.callbackify.project');
```
I don't think we need these properties - they complicate the PR and with destructuring syntax and a uniform convention none of the benefits it would bring to `util.promisify` apply here IMO.

For example, it is required for `fs.exists` to work in `promisify` but there is no equivalent here.

refack replied:
Came to same conclusion. Gone in last commit.

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+
+  // we have two way to convert the returned value into multiple arguments
+  // either splat it or project some of it's keys
+  const toSplat = options.toSplat;
```
I don't think we should support this - with `promisify` we don't really have a choice for things like `fs.exists` but here destructuring syntax should be sufficient.

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+  const toSplat = options.toSplat;
+  const project = options.project;
+
+  const fn = function(...args) {
```
Using `...args` `args.pop` and `call` with `...args` is expensive here.

I would prefer to see `arguments` used, `apply` instead of call, and an inline slice instead of `.pop` for performance like bluebird does.

refack replied:
Since this will go on top of TF+I, is this still a major issue?

jasnell replied:
Just noting to keep it in mind, it's not yet clear *when* we'll be enabling TF+I. It's entirely possible that this could land before that but hard to say for sure.

refack replied:
I'll cook up some benchmarks (also look for DEOPTS)...

benjamingr replied:
If you run this on top of TF+I (by checking out the appropriate branch) and the results show that these optimizations are meaningless - then we can drop them. I suspect they still affect execution speed on TF+I.

addaleax replied:
@benjamingr Fwiw, there have been PRs like https://github.com/nodejs/node/pull/12183 opened by V8 people. I’m good with using rest parameters here, at least for this PR.

benjamingr replied:
Interesting, @hashseed care to weigh in?

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+    }
+    var p = orig.call(this, ...args);
+    if (typeof options.transformer === 'function') {
+      p = options.transformer(p);
```
similar comment about not needing a transformer.

hashseed replied:
Rest parameters are at least as fast as arguments object, and avoids many performance cliffs. Spread is just as fast as `Function.prototype.apply`. If I remember correct this applies to both Crankshaft and Turbofan.

@psmarshall and @bmeurer has more details though.

hashseed replied:
In any case, the runtime of this function is likely to be dominated by the chaining of the promise that follows. I wouldn't micro-optimize this just yet, and opt for the cleaner syntax.

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+      p = options.transformer(p);
+    }
+    p = p.then((ret) => {
+      if (toSplat) {
```
You _have_ to call `cb` inside a `process.nextTick` in case it throws - you can't just abort the process or emit an event since it has to work with domains.

This is what bluebird and other libraries do with their implementations and it's the most reasonable thing to do.

I would also avoid all the `if/else` here and the projection - which users can just do in a second `then`.

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+        cb(null, ret);
+      }
+    }, (rej) => {
+      // !rej guard inspired by BlueBird https://goo.gl/t5IS6M
```
At bluebird, this improves debugging experience. In Node this is not something we can afford - converting non-errors into errors is useful for a library but it's not something the platform should do.

refack replied:
Ref: https://github.com/nodejs/node/pull/12712#discussion_r113844782

benjamingr replied:
I saw that, but we can't assume that if someone threw a falsey value it should be converted to an error. Bluebird does this to overcome the lack of stack trace but node should not make that decision for users. As a platform we should provide a more careful API.

not-an-aardvark replied:
To clarify, I didn't suggest doing this so that errors would have a stack trace -- I suggested it because `null` is a special error value in callbacks which means "no error occurred". If we don't error-wrap, there is no way for a callback consumer to distinguish between "the promise rejected with `null`" and "the promise fulfilled with `undefined`".

benjamingr commented on lib/util.js:
```diff
@@ -1057,3 +1057,72 @@ exports._exceptionWithHostPort = function(err,
+        rej = newRej;
+      }
+      cb(rej);
+    }).catch((e) => process.emit('uncaughtException', e));
```
Ditto, this catch is incorrect, and the error must be emitted in `nextTick`, otherwise this breaks post-mortem debugging.

refack replied:
I'm looking into this (and the `nextTick`)
Ref: https://github.com/nodejs/node/pull/12712#discussion_r113928280

benjamingr replied:
@refack the problem isn't that catch happens on `nextTick`, it's that if you abort on unhandled exception the error would be thrown properly and the heap would not be possibly corrupted. 

(Also, `catch` happens after nextTick, not during)

refack replied:
So you want an "expulsion", callback throwing style? Makes sense.

refack commented on lib/util.js:
```diff
@@ -1110,28 +1108,21 @@ function callbackify(orig) {
       cb(rej);
-    }).catch((e) => process.emit('uncaughtException', e))
+    }).catch((e) => process.emit('uncaughtException', e));
+    return p;
```
Should we return the promise?

refack commented:
Last commit goes full-on with the options.
For your consideration...

refack commented on lib/util.js:
```diff
@@ -1057,3 +1057,75 @@ exports._exceptionWithHostPort = function(err,
+      }
+      cb(rej);
+    }).catch((e) => process.emit('uncaughtException', e));
+    return p;
```
Should we return the promise?

benjamingr replied:
No, we should return `undefined`

refack replied:
Reason?

benjamingr replied:
It gives the user a false sense that the promise is actually somehow related to the callback's execution and creates a confusing API in which the user might expect the callback throwing to reject the promise.

This is why we switched from returning the promise to returning `undefined` in bluebird.

refack replied:
I'm going to document that

benjamingr replied:
Basically, people were doing things like the following with BB:

```js
await promise.asCallback((err, data) => {
  console.log("2");
});
console.log("1");
```
And expected 2, 1 to log (and got 1,2), not to mention people thinking that if they nest callbacks it will somehow magically wait for them.

refack replied:
Documented.
_But_ added an option to opt-in to such behaviour 😉

not-an-aardvark commented:
> Last commit goes full-on with the options.
For your consideration...

I don't think this is necessary either, for the same reasons as https://github.com/nodejs/node/pull/12712#issuecomment-297904232. I think we should just remove these options and the related functionality.

sam-github commented:
No docs.... no comment (how could I if I don't know what its supposed to do?).

No node API returns promises, does it? What is the API for? (see above.... "no docs").

addaleax commented on lib/util.js:
```diff
@@ -1057,3 +1057,80 @@ exports._exceptionWithHostPort = function(err,
+    const cb = args.pop();
+    if (typeof cb !== 'function') {
+      throw new TypeError('The first argument to the callbackified function ' +
+        'must be a callback style function');
```
`last argument`?

refack replied:
Fixed.

addaleax commented on lib/util.js:
```diff
@@ -1057,3 +1057,80 @@ exports._exceptionWithHostPort = function(err,
+      p = options.transformer(p);
+    }
+    p = p.then((ret) => process.nextTick(onFulfilled, {ret, options, cb}),
+               (rej) => process.nextTick(onRejected, {rej, cb}));
```
Is there any reason for creating objects here? Just passing them as 2 or 3 extra arguments to `nextTick` should be fine

refack replied:
I had something else in mind, but now you're obviously right. Fixed.
(`p.then((ret) =>{ret, options, cb}).then(onFulfilled)`)

refack commented:
While the `options` is still in debate, I added `option.catch` allowing the user to tie-in the callback to the underlying promise's life cycles and, hance also returning the promise in this case.

refack commented:
With latest comments incorporated IMHO two `options` have a meaningful use case
1. `project` - when the expected consumer is a user of a mixed API and expects uniform callback signatures (where the wrapped `awaitable` is a part of an API made by 3rd party)
2. `catch` - opt-in to incorporate the callback into the lifecycle of the underlying `awaitable`, i.e. having
    ```
    await promise.asCallback((err, data) => {
       console.log("1");
    });
    console.log("2");
    ```
     print `1, 2`

benjamingr commented:
@refack `catch` should be on my default (otherwise what, no errors?) and `project` can just be something you call on the function later.

I think both of these still complicate the PR beyond scope, and may be added later if we _really_ want them for some reason.

refack commented:
@benjamingr I think you mean `catch` should be off (cb runs on next tick), when it's on, the cb returns a promise and errors cought into the Promise life-cycle.
As for `project`, IMHO we are building a `util`, something devs use to make uniform APIs I don't assume they are all the direct consumers.

For example:
#### MyAPI.js
```js
const {GetA, GetB} require('GetsStuffAsPromise);
function GetC(id, cb) {
  ...
  return cb(err, C, idx, num)
}

module.export = {
  MyGetA: util.calbackify(GetA, {project:'*'}),
  MyGetB: util.calbackify(GetA, {project:['Obj', 'idx', 'num']}),
  MyGetC: GetC
}
```

not-an-aardvark commented:
I agree with @benjamingr -- I don't think there is enough demonstrated need for options on this function. If someone wants to expose a consistent API then they can wrap the resulting function themselves -- we can't account for all of the return formats that someone might need anyway. For now, I think `util.callbackify` should just accept a single function without options, customization symbols, or anything else.

benjamingr commented:
>  I think you mean catch should be off (cb runs on next tick), when it's on, the "callbackified" wrapper returns a promise and errors are caught into the Promise life-cycle.

Right, trapping the error in callbackify is a no-go, the callback has to run globally like any other callback so that errors cause the process to exit since it's in an unclean state.

refack commented:
@benjamingr @not-an-aardvark removed all frills... PTAL

not-an-aardvark commented on lib/util.js:
```diff
@@ -1057,3 +1057,55 @@ exports._exceptionWithHostPort = function(err,
+}
+
+
+function _cbfy_argsAssert(args) {
```
Can we rename this function to something like `assertLastArgFunction`? The current name is not very descriptive.

refack replied:
Done

not-an-aardvark commented on lib/util.js:
```diff
@@ -1057,3 +1057,55 @@ exports._exceptionWithHostPort = function(err,
+                 'callback style function';
+
+
+function _cbfy_onRejected(rej, cb) {
```
Can we rename this function to something like `function callbackifyOnRejected(reason, cb)`? The current function signature is sort of hard to read.

refack replied:
Sure.

not-an-aardvark commented on lib/util.js:
```diff
@@ -1057,3 +1057,55 @@ exports._exceptionWithHostPort = function(err,
+    newRej.cause = rej;
+    rej = newRej;
+  }
+  return cb(rej);
```
It looks like you removed the `process.nextTick` call, so this behavior will be incorrect if the callback function throws an error.

refack replied:
Opps, wrong remove.

refack commented:
@not-an-aardvark I wanted for us to converge on the signature. Now to tests 🏃 ...

addaleax commented:
ping @refack

refack commented:
> ping @refack

@addaleax thanks!

eljefedelrodeodeljefe commented:
Can we ship this rather soon? ping @refack and @benjamingr @not-an-aardvark for review

mcollina commented on lib/util.js:
```diff
@@ -1058,4 +1058,58 @@ exports._exceptionWithHostPort = function(err,
+  // occurred". We error-wrap so the callback consumer can distinguish between
+  // "the promise rejected with null" or "the promise fulfilled with undefined".
+  if (!reason) {
+    const newRej = new Error(reason + '');
```
What if `reason` is an `Error` object?

addaleax replied:
@mcollina This is in the `if (!reason)` branch, so I am not sure what exactly you mean by that; could you be more specific?

mcollina replied:
ouch, sorry. Forget about my comment.

addaleax commented:
> Can we ship this rather soon?

This is currently still lacking tests and documentation, so it can’t land in the next few days. Maybe somebody wants to pick this up?

cjihrig commented:
> Maybe somebody wants to pick this up?

@refack are you still working on, or planning to work on, this?

refack commented:
I really mean to but keep getting distracted. If someone can help with some test cases, that would be great help.

refack commented:
linuxone quick CI: https://ci.nodejs.org/job/node-test-commit-linuxone/6357/

refack commented:
Green linuxone CI: https://ci.nodejs.org/job/node-test-commit-linuxone/6360/
To whom it may concern I think this it ready. PTAL.
More test cases are welcome 👐

not-an-aardvark commented on doc/api/util.md:
```diff
@@ -10,6 +10,42 @@ module developers as well. It can be accessed using:
+
+*Note*: Like with most "callback style" function, the callback is executed in an
+async context (having a limited stacktrace), and if the callback throws, the
+process will emit an `uncaughtException` event and if not handles will exit.
```
Nit: I think this should say "if not *handled*".

not-an-aardvark commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,49 @@
+  const cbFn = callbackify(fn);
+
+  cbFn(common.mustCall((err, ret) => {
+    if (err) assert.ifError(err);
```
I think using `if (err)` here is redundant because `assert.ifError` already is a no-op if `err` is falsy.

refack replied:
Ack

not-an-aardvark commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,49 @@
+  cbFn(common.mustCall((err, ret) => {
+    if (err) assert.ifError(err);
+    common.mustNotCall('should not reach here')();
+  }));
```
Can you add a test to make sure an rejection reason of `null` is wrapped in an `Error`?

refack replied:
Did some. Will add all as soon as...

refack replied:
Done

not-an-aardvark commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+  // occurred". We error-wrap so the callback consumer can distinguish between
+  // "the promise rejected with null" or "the promise fulfilled with undefined".
+  if (!reason) {
+    const newRej = new Error(reason + '');
```
This will throw if a Promise is rejected with a `Symbol` value. It would be better to use `new Error(String(reason))`.

benjamingr replied:
@not-an-aardvark nice catch!

not-an-aardvark commented on doc/api/util.md:
```diff
@@ -10,6 +10,42 @@ module developers as well. It can be accessed using:
+
+cbFn((err, ret) => {
+  if (err) throw err;
+  console.log(ret);
```
It might be good to add a comment somewhere indicating that `err` will be the rejection reason if the Promise resolves, and `ret` will be the value of the Promise fulfills.

mcollina commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+    const cb = assertLastArgIsFunction(args);
+    asyncEndpoint.call(this, ...args)
+      .then((ret) => process.nextTick(cb, null, ret),
+            (rej) => process.nextTick(callbackifyOnRejected, rej, cb)
```
I think we should add a test about exception handling and `process.nextTick`. Specifically, I would like to see that if there is a `throw new Error('kaboom')`  in the callback that error ends up in `uncaughtException`.

thefourtheye replied:
Why `cb` is passed when it is actually a rejection?

mcollina commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+    asyncEndpoint.call(this, ...args)
+      .then((ret) => process.nextTick(cb, null, ret),
+            (rej) => process.nextTick(callbackifyOnRejected, rej, cb)
+      );
```
Can we add a test that verifies that if `cb()` throws, the error ends up in `'uncaughtException'`? This is different from the test below, as it verifies that a rejection causes the same effect.

benjamingr commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+
+
+function assertLastArgIsFunction(args) {
+  const cb = args.pop();
```
Not a fan of calling the method `assert` and then mutating `args` in it.

mcollina replied:
I think this functionality can be moved to calling function.

benjamingr commented:
The latest changes look good to me, if you address the feedback by @not-an-aardvark I'll approve.

Thanks for the work and bearing with the churn of all the discussions with us @refack  <3

addaleax commented on doc/api/util.md:
```diff
@@ -10,6 +10,42 @@ module developers as well. It can be accessed using:
+async function fn() {
+  return await Promise.resolve('hello world');
+}
+const cbFn = util.callbackify(fn);
```
nit: I’d just write it out as `callbackFunction`, clarity > brevity :)

addaleax commented on doc/api/util.md:
````diff
@@ -10,6 +10,42 @@ module developers as well. It can be accessed using:
+});
+```
+
+will output something like:
````
nit: output → print, that’s the term we usually use in our documentation

addaleax commented on doc/api/util.md:
````diff
@@ -10,6 +10,42 @@ module developers as well. It can be accessed using:
+hello world
+```
+
+*Note*: Like with most "callback style" function, the callback is executed in an
````
typo: `function` → `functions`. Again, I’d drop the quotes.

addaleax commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+ * @function callbackify
+ * @param {function(): Promise} asyncEndpoint
+ * @public
+ */
```
I know other collaborators have asked you to remove these blocks elsewhere, and this one in particular doesn’t really seem to provide anything useful

addaleax commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+}
+
+exports.callbackify = callbackify;
+// End Callbackify
```
We don’t do this elsewhere either

addaleax commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,49 @@
+
+
+{
+  const sentinel = 'hello world';
```
`sentinel`?

refack replied:
https://en.wikipedia.org/wiki/Sentinel_value adopted in TDD (at least back in the day) as an Object you use to guard with...

addaleax commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,49 @@
+  const sentinel = new Error('hello world 3');
+  process.once('uncaughtException', common.mustCall((err) => {
+    assert.strictEqual(err, sentinel);
+  }));
```
I would move this into a separate test file, otherwise the `uncaughtException` handler here might interfere with the rest of the test

refack replied:
Ack. already experienced that.

refack commented:
Tracking new test cases
- [X] A rejection reason of null is wrapped in an Error when passed to the callback
- [ ] If the callbackified function is called with any arguments, they get passed through to the promise-returning function
- [ ] The promise-returning function is called with the same this value as the callbackified version
- [ ] Nothing crashes when the Promise rejection reason is a Symbol
- [ ] The err argument has a cause property when the rejection reason is falsy (do we actually want to do this? I'm not convinced it's necessary to have a property that only sometimes exists and is always falsy, but if we're setting the property we should have a test for it.)

mcollina commented on test/fixtures/uncaught-exceptions/callbackify1.js:
```diff
@@ -0,0 +1,19 @@
+  const sentinel = new Error(__filename);
+
+  async function fn3 () {
+    return await Promise.reject(sentinel);
```
there is no need to use `Promise.reject()` here.

benjamingr replied:
What do you mean?

refack replied:
Just throw something.
Done.

mcollina commented on test/fixtures/uncaught-exceptions/callbackify1.js:
```diff
@@ -0,0 +1,19 @@
+  const cbFn = callbackify(fn3);
+
+  cbFn((err, ret) => {
+    throw err;
```
just throw a new `Error` here.

refack replied:
Yep, so we don't need the `sentinel` either

benjamingr commented on doc/api/util.md:
```diff
@@ -10,6 +10,45 @@ module developers as well. It can be accessed using:
+* `original` {Function} An `async` function
+* Returns: {Function} a callback style function
+
+Takes an `async` function (or a function that returns a Promise) and returns a
```
I would probably start this with "This method converts promise returning APIs to callback taking ones" or something super explicit at the start.

refack replied:
Ack

benjamingr commented on doc/api/util.md:
````diff
@@ -10,6 +10,45 @@ module developers as well. It can be accessed using:
+```
+
+*Note*: Like with most callback style functions, the callback is executed in an
+async context (having a limited stacktrace), and if the callback throws the
````
Note that with the debugger attached the stack trace isn't limited (async stack traces). Although I think it's a fair statement here.

benjamingr commented on lib/util.js:
```diff
@@ -1045,3 +1045,51 @@ process.versions[exports.inspect.custom] =
+    }
+    const cb = maybeCb.bind(this);
+    original.call(this, ...args)
+      .then((ret) => process.nextTick(cb, null, ret),
```
It's worth mentioning that `process.nextTick` is here so the `then` does not handle the errors - but I'm fine with it

refack replied:
Adding a comment (as you saw there is a *Note*: in the docs)

benjamingr commented on lib/util.js:
```diff
@@ -1045,3 +1045,51 @@ process.versions[exports.inspect.custom] =
+        'last argument',
+        'function');
+    }
+    const cb = maybeCb.bind(this);
```
Do we really not have a way to pass context to nextTick without binding or a closure :(? 

Although bind is fast now, so I guess it doesn't matter.

refack replied:
I'm not super in love with it either since it hides the function's origin in the debugger

refack commented:
Re: https://github.com/nodejs/node/pull/12712#discussion_r120115826
`const cb = maybeCb.bind(this);`
Does anyone have a better idea how to preserve the context for the callback?

Other then that, AFAICT I addressed all comments. Me. Add. Words. Comments. Doc. Good? 🦍 
PTAL.

refack commented:
CI: https://ci.nodejs.org/job/node-test-commit/10367/

benjamingr commented:
> Does anyone have a better idea how to preserve the context for the callback?

Well, an anonymous function also works, but seeing nextTick doesn't take a context parameter I'm not sure we have a choice.

refack commented:
> Well, an anonymous function also works

To close over `this`? 🤔 It will preserve the functions' code origin.
I'll benchmark.

refack commented:
Also PTAL at 11a3a03980

benjamingr commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
+
+
+function callbackifyOnRejected(reason, cb) {
+  // !reason guard inspired by BlueBird https://goo.gl/t5IS6M
```
Yes, it's bluebird

benjamingr commented on lib/util.js:
```diff
@@ -1045,3 +1045,58 @@ process.versions[exports.inspect.custom] =
 exports.promisify = internalUtil.promisify;
+
+// Callbackify section
+const CBFY_ERR = 'The last argument to the callbackified function must be a ' +
```
constants are typically `kCallbackifyError`

jasnell commented on doc/api/util.md:
````diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+```
+
+*Note*:
+* Like with most callback style functions, the callback is executed in an
````
No new line after `*Note*:`

jasnell commented on doc/api/util.md:
````diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+});
+```
+
+will print something like:
````
`Will print:`

jasnell commented on doc/api/util.md:
```diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+async context (having a limited stacktrace), and if the callback throws the
+process will emit an `uncaughtException` event, and if not handled, will exit.
+* Since `null` has a special meaning as the first argument to a callback, if a
+wrapped function rejects a `Promise` with a "falsy" value as a reason, we wrap
```
Please avoid the use of pronouns like `we`

refack replied:
Ack.

cjihrig replied:
I think you can drop the quotes around falsy.

jasnell commented on doc/api/util.md:
````diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+});
+```
+
+
````
Unnecessary extra blank lines

refack replied:
Ack

jasnell commented on lib/util.js:
```diff
@@ -1045,3 +1045,54 @@ process.versions[exports.inspect.custom] =
+  // occurred", we error-wrap so the callback consumer can distinguish between
+  // "the promise rejected with null" or "the promise fulfilled with undefined".
+  if (!reason) {
+    const newRej = new Error(String(reason));
```
Please use `internal/errors` here too. A special error code can be assigned for this.

refack replied:
Added `NULL_REJECTION`

cjihrig commented on doc/api/util.md:
```diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+
+*Note*:
+* Like with most callback style functions, the callback is executed in an
+async context (having a limited stacktrace), and if the callback throws the
```
Can you stop the sentence after stacktrace. Then, start a new sentence.

cjihrig commented on doc/api/util.md:
```diff
@@ -10,6 +10,64 @@ module developers as well. It can be accessed using:
+const callbackFunction = util.callbackify(fn);
+
+callbackFunction((err, ret) => {
+  err && err.cause && err.cause === null;  // true
```
What exactly are you trying to show here?

cjihrig commented on test/fixtures/uncaught-exceptions/callbackify2.js:
```diff
@@ -0,0 +1,20 @@
+  const sentinel = new Error(__filename);
+  process.once('uncaughtException',(err) => {
+    assert.strictEqual(err, sentinel);
+    console.log(err.message);
```
Can the `console.log()` be removed if it isn't part of the test.

addaleax replied:
In that case, leaving a comment that it’s part of the test might be a good idea :)

refack replied:
Ack.

cjihrig commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,179 @@
+'use strict';
+const common = require('../common');
+
```
Unnecessary blank line.

cjihrig commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,179 @@
+}
+
+{
+  // 3. args work
```
Could you make this, and the other similar comments more meaningful, or drop them completely.

addaleax commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+*Note*:
+
+* Like with most callback style functions, the callback is executed in an
+async context (i.e. having a limited stacktrace). If the callback throws, the
```
having → has, right?

refack replied:
I'm not sure, I don't know grammar rules, just intuition.
If the sentence was "the callback has a limited stacktrace" yes.
But I think it's "the callback is executed in a context having a limited stacktrace".

BridgeAR replied:
Having should be correct.

cjihrig replied:
Can you shorten the first sentence to:


The callback is executed asynchronously, and will have a limited stack trace.

refack replied:
Ack

addaleax commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+
+* Like with most callback style functions, the callback is executed in an
+async context (i.e. having a limited stacktrace). If the callback throws, the
+process will emit an `uncaughtException` event, and if not handled will exit.
```
Can you link to the uncaughtException docs?

refack replied:
ACk

addaleax commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+callbackFunction((err, ret) => {
+  // When the Promise was rejected with `null` it is wrapped with an Error and
+  // the original value is stored in `reason`.
+  err && ('reason' in err) && err.reason === null;  // true
```
I think in this case `err.hasOwnProperty` might be better than `in`, it’s a bit more idiomatic and – I think – more performant because there’s no reason to walk the prototype chain.

refack replied:
Ack

BridgeAR replied:
`in` should be a tiny bit [faster](https://jsperf.com/hasownproperty-vs-in-vs-undefined/128) than `hasOwnProperty`.

refack replied:
Since it's an example snippet I'm pro being more idiomatic. (although I'm old and for me `'x' in obj` __is__ the idiom, but I agree times have changed)

addaleax commented on lib/util.js:
```diff
@@ -1045,3 +1045,55 @@ process.versions[exports.inspect.custom] =
+    Reflect.apply(original, context, args)
+      .then((ret) => process.nextTick(cb, null, ret),
+            (rej) => process.nextTick(callbackifyOnRejected, rej, cb)
+      );
```
does the `);` fit on the previous line? :p

Also, I don’t think we have a test in here that checks that `original()` can return a plain thenable, and people are likely going to rely on that if we allow it (which seems perfectly fine to me).

refack replied:
Ack on `)`
As for test: easy way to create a "Thenable"?

addaleax replied:
@refack I was thinking of a plain object with a single `then(onFulfill, onReject)` method, i.e. the bare minimum in order to work fine with other Promise implementations

addaleax commented on test/fixtures/uncaught-exceptions/callbackify2.js:
```diff
@@ -0,0 +1,20 @@
+
+{
+  const sentinel = new Error(__filename);
+  process.once('uncaughtException',(err) => {
```
space after the comma?

refack replied:
Ohh the linter ignore the fixtures.
Ack.

addaleax commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,188 @@
+const { join } = require('path');
+const { execFile } = require('child_process');
+const fixtureDir = join(common.fixturesDir, 'uncaught-exceptions');
+const sentinelValues = [
```
`sentinel`?

refack replied:
We had that conversation...
It's TDD speak...
IMHO it's better than `values` but you can 👍 / 👎

addaleax replied:
I would prefer `values`, I think.

refack replied:
Ack

addaleax commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,188 @@
+      if (err instanceof Error) {
+        if ('reason' in err) {
+          assert.strictEqual(err.code, 'NULL_REJECTION');
+          assert.strictEqual(err.reason, sentinel);
```
Maybe just leave `null` off the list, and have separate tests.

If not, I think this block is supposed to have an `assert.strictEqual(sentinel, null);` check?

refack replied:
reason could be any falsy value (as stated in the docs, and the check is `if(!reason)`

addaleax replied:
Maybe at least check `assert(!sentinel);` / `assert(!value);` then? ;)

refack replied:
Ack (and `sentinel` -> `value` virtue of WebStorm's Refactor:Rename)

refack commented:
linuxone quick CI: https://ci.nodejs.org/job/node-test-commit-linuxone/6466/

refack commented:
https://ci.nodejs.org/job/node-test-commit-linuxone/6467/

refack commented:
@addaleax I had an idea for a follow up PR to turn this to more of an `promisify`<sup>-1</sup> keep the `original` somewhere and just return it, i.e. `callbackify(promisify(fn)) === fn` and `promisify(callbackify(fn)) === fn`

addaleax commented:
@refack Let’s keep that for another discussion; I would be really careful about that, for example (as I’ve mentioned above) the original Promise-returning function doesn’t need to return *native* Promises, but the one returned by `promisify()`always does.

not-an-aardvark commented on lib/util.js:
```diff
@@ -1045,3 +1045,54 @@ process.versions[exports.inspect.custom] =
+  // occurred", we error-wrap so the callback consumer can distinguish between
+  // "the promise rejected with null" or "the promise fulfilled with undefined".
+  if (!reason) {
+    const newReason = new errors.Error('NULL_REJECTION');
```
It might be confusing to have an error that says "The Promise rejected with null" if the Promise rejected with some other falsy value.

refack replied:
Ack.
Changed to `'FALSE_REJECTION', 'Promise was rejected with falsy value'`

not-an-aardvark commented on test/parallel/test-util-callbackify.js:
```diff
@@ -0,0 +1,192 @@
+  // Test that callback that throws emits an `uncaughtException` event
+  const fixture = join(fixtureDir, 'callbackify1.js');
+  execFile(
+    process.argv[0],
```
Nit: Maybe `process.execPath` would be clearer than `process.argv[0]`, but it doesn't matter much either way.

refack replied:
Ack

not-an-aardvark commented:
> Does anyone have a better idea how to preserve the context for the callback?

Do we actually need to call the callback with a `this` value? As far as I know, callbacks generally don't have any specified `this` value. (Analogously, functions in `.then` arguments are explicitly forbidden from being called with a `this` value.)

TimothyGu commented on doc/api/util.md:
```diff
@@ -10,6 +10,68 @@ module developers as well. It can be accessed using:
+* Returns: {Function} a callback style function
+
+This method converts Promise returning API endpoints to ones that use callbacks.
+`callbackify` takes an `async function` (or a function that returns a Promise)
```
Nittest of nits: MDN uses "`async` function" backticks only around `async`, which is IMO more consistent since we don't use `function`.

refack replied:
Ack.

TimothyGu commented on doc/api/util.md:
````diff
@@ -10,6 +10,68 @@ module developers as well. It can be accessed using:
+is wrapped in an `Error` with the original value stored in a field named
+`reason`.
+
+```js
````
Indent this code block by 2 spaces since it clarifies the second point.

refack replied:
Ack

TimothyGu commented on lib/util.js:
```diff
@@ -1045,3 +1045,54 @@ process.versions[exports.inspect.custom] =
+        'function');
+    }
+    const context = this;
+    function cb() { Reflect.apply(maybeCb, context, arguments); }
```
Why not use an arrow function, which would render `context` unnecessary?

BridgeAR replied:
Is there a performance benefit by using Reflect?

refack replied:
@TimothyGu Just went with an intuition that a named function & aliasing `this` with `context` would be more "readable"
Feel free to 👍 / 👎

refack replied:
@BridgeAR Re: https://github.com/nodejs/node/issues/12956
```
> function cb(arg) {console.log(this); console.log(arg);}
undefined
> cb.apply = 5
5
> cb.apply({}, 1)
TypeError: cb.apply is not a function
    at repl:1:4
    at ContextifyScript.Script.runInThisContext (vm.js:44:33)
    at REPLServer.defaultEval (repl.js:239:29)
    at bound (domain.js:301:14)
    at REPLServer.runBound [as eval] (domain.js:314:12)
    at REPLServer.onLine (repl.js:433:10)
    at emitOne (events.js:120:20)
    at REPLServer.emit (events.js:210:7)
    at REPLServer.Interface._onLine (readline.js:278:10)
    at REPLServer.Interface._line (readline.js:625:8)
>
```
AFAIK there is no penalty.

not-an-aardvark replied:
> Just went with an intuition that a named function & aliasing `this` with context would be more "readable"

I personally think an arrow function would be more readable, but I don't feel strongly about this either way.

gibfahn replied:
+1 to an arrow function for readability.

refack replied:
Ack. Arrow function it is.
@not-an-aardvark Now the `this` binding is for free, so IMHO it's better then `undefined` binding. + / - ?

addaleax commented on lib/internal/errors.js:
```diff
@@ -164,7 +164,7 @@ E('ERR_SOCKET_BAD_PORT', 'Port should be > 0 and < 65536');
 E('ERR_V8BREAKITERATOR', 'full ICU data not installed. ' +
   'See https://github.com/nodejs/node/wiki/Intl');
-E('NULL_REJECTION', 'Promise was rejected with "null"');
+E('FALSE_REJECTION', 'Promise was rejected with falsy value');
```
The message change is okay, but I liked `NULL_REJECTION` better as a code; `FALSE_REJECTION` sounds like “this is not really a rejection”, even if it totally is one.

refack replied:
🤔 hhhhmmmm...
`'FALSY_VALUE_REJECTION'`?

refack replied:
Went with `FALSY_VALUE_REJECTION`

refack commented:
https://ci.nodejs.org/job/node-test-commit-linuxone/6476/

addaleax commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+* `original` {Function} An `async` function
+* Returns: {Function} a callback style function
+
+This method converts Promise returning API endpoints to ones that use callbacks.
```
“API endpoints”? It takes functions, independent of context … I realize that would be a bit of duplication with the next sentence, but I’d rather have duplication than confusion (also, I’m quite sure it’s `Promise-returning` if you use it as an attribute)

refack replied:
I open to change, but it's based on @benjamingr 's phrasing https://github.com/nodejs/node/pull/12712#discussion_r120115166
I think the idea is to give the motivation. Maybe:
`This utility is provided to help in adapting Promise-returning APIs to callback style.`

cjihrig replied:
Instead of API endpoints, can you just say functions? API endpoints may cause people to think of things like HTTP endpoints.

gibfahn replied:
Do we even need this sentence? As @addaleax said, it's basically covered by the next sentence. I think:

```
Takes an `async` function (or a function that returns a Promise)
and returns a function following the Node.js error first callback style,
```

is a perfect introduction to this function.

refack replied:
Removed first sentence
@benjamingr PTAL.

refack commented:
Sanity check https://ci.nodejs.org/job/node-test-commit-linuxone/6476/

BridgeAR commented on lib/util.js:
```diff
@@ -1045,3 +1045,54 @@ process.versions[exports.inspect.custom] =
+  // the promise is actually somehow related to the callback's execution
+  // and that the callback throwing will reject the promise.
+  function callbackified(...args) {
+    const maybeCb = args.pop();
```
Nit: moving the maybeCb out of the args seems nicer to me.
```js
function callbackified(maybeCb, ...args) {
```

TimothyGu replied:
`maybeCb` must be the last parameter, so that won't work. `(...args, maybeCb)` is not valid syntax either.

BridgeAR replied:
Oh, ignore my comment, I somehow didn't see that it wasn't shift...

refack commented:
Quick CI: https://ci.nodejs.org/job/node-test-commit-linuxone/6533/

refack commented:
Full CI: https://ci.nodejs.org/job/node-test-pull-request/8590/

refack commented:
@addaleax @cjihrig PTAL
/cc @benjamingr (comment on https://github.com/nodejs/node/pull/12712#discussion_r120862852)

cjihrig commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+
+This method converts Promise returning API endpoints to ones that use callbacks.
+`callbackify` takes an `async` function (or a function that returns a Promise)
+and returns a function following the common Node.js callback style, i.e. taking
```
The part of the sentence starting with "following the common Node.js" could probably just be "following the Node.js error first callback style."

refack replied:
Ack

cjihrig commented on doc/api/util.md:
````diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+});
+```
+
+Will print something like:
````
I think @jasnell commented on this, but please change to "Will print:"

refack replied:
Ack.

cjihrig commented on doc/api/util.md:
```diff
@@ -10,6 +10,67 @@ module developers as well. It can be accessed using:
+  callbackFunction((err, ret) => {
+    // When the Promise was rejected with `null` it is wrapped with an Error and
+    // the original value is stored in `reason`.
+    err && err.hasOwnProperty('reason') && err.reason === null;  // true
```
Is the `hasOwnProperty()` check really necessary?

refack replied:
The thought was to be very explicit since this is an example.
(Discussed in previous review cycles)

cjihrig commented on test/fixtures/uncaught-exceptions/callbackify1.js:
```diff
@@ -0,0 +1,15 @@
+const { callbackify } = require('util');
+
+{
+  async function fn3() { }
```
Why `fn3`?

refack replied:
renamed to `fn`

cjihrig commented on test/fixtures/uncaught-exceptions/callbackify2.js:
```diff
@@ -0,0 +1,22 @@
+    console.log(err.message);
+  });
+
+  async function fn3() {
```
Same comment about `fn3`.

refack replied:
renamed to `fn`

refack commented:
Quick sanity: https://ci.nodejs.org/job/node-test-commit-linuxone/6543/

not-an-aardvark commented:
> Quick sanity: https://ci.nodejs.org/job/node-test-commit-linuxone/6543/

Can we make sure to run the full CI at some point before this gets landed? I don't expect that the behavior would be different on different platforms, but it would be good to make sure.

refack commented:
>> Quick sanity: https://ci.nodejs.org/job/node-test-commit-linuxone/6543/
>
> Can we make sure to run the full CI at some point before this gets landed? I don't expect that the behavior would be different on different platforms, but it would be good to make sure.

Sure: https://ci.nodejs.org/job/node-test-pull-request/8596/

@cjihrig PTAL.

refack commented:
Landed in af3aa682ac

refack commented:
Thanks everybody for all the feedback and insights 🎆 🍾

jasnell commented on lib/internal/errors.js:
```diff
@@ -164,6 +164,7 @@ E('ERR_SOCKET_BAD_PORT', 'Port should be > 0 and < 65536');
 E('ERR_SOCKET_DGRAM_NOT_RUNNING', 'Not running');
 E('ERR_V8BREAKITERATOR', 'full ICU data not installed. ' +
   'See https://github.com/nodejs/node/wiki/Intl');
+E('FALSY_VALUE_REJECTION', 'Promise was rejected with falsy value');
```
dang... I should have caught this one before it landed... this needs the `ERR_` prefix on it. That should be fixed before this goes out in a release.

ping @refack @addaleax

benjamingr commented:
Thanks a lot @refack! good job :)

addaleax commented:
@refack Can you backport (together with 873e2f270f)?
